[clang] Replace uses of CreatePointerBitCastOrAddrSpaceCast (NFC)

JOE1994 opened the pull request:
With opaque pointers, CreatePointerBitCastOrAddrSpaceCast is same as CreateAddrSpaceCast. Replace or remove uses of CreatePointerBitCastOrAddrSpaceCast.

Opaque pointer cleanup effort.

JOE1994 committed:
[clang] Replace uses of CreatePointerBitCastOrAddrSpaceCast (NFC)

With opaque pointers, CreatePointerBitCastOrAddrSpaceCast is same as CreateAddrSpaceCast.
Replace or remove uses of CreatePointerBitCastOrAddrSpaceCast.

Opaque pointer cleanup effort.

JOE1994 commented:
Applied git clang-format.

JOE1994 commented:
No new regressions with `ninja check-clang`

nikic commented on llvm/include/llvm/IR/IRBuilder.h:
```diff
@@ -2138,6 +2138,8 @@ class IRBuilderBase {
     return Insert(CastInst::CreatePointerCast(V, DestTy), Name);
   }
 
+  // With opaque pointers enabled, this is same as CreateAddressSpaceCast.
```
All the IRBuilder cast methods are no-ops if old type == new type.

s-barannikov replied:
I'll try to rephrase @arsenm comment.
If the types (= address spaces) are the same, CreatePointerBitCastOrAddrSpaceCast does nothing, while CreateAddressSpaceCast aborts. That is, they are equivalent iff the types are different.

JOE1994 replied:
> If the types (= address spaces) are the same, CreatePointerBitCastOrAddrSpaceCast does nothing, while CreateAddressSpaceCast aborts. That is, they are equivalent iff the types are different.

Thank you for the clarification!

So `CreateAddrSpaceCast` and `CreatePointerBitCastOrAddrSpaceCast` aren't strict equivalents,
but it still seems that uses of `CreatePointerBitCastOrAddrSpaceCast` can be substituted with `CreateAddrSpaceCast`.

JOE1994 replied:
```suggestion
  // With opaque pointers enabled, this can be substituted with CreateAddrSpaceCast.
```

s-barannikov commented on clang/lib/CodeGen/ItaniumCXXABI.cpp:
```diff
@@ -3088,9 +3088,6 @@ void ItaniumCXXABI::EmitThreadLocalInitFuncs(
       Val = Builder.CreateAlignedLoad(Var->getValueType(), Val, Align);
     }
-    if (Val->getType() != Wrapper->getReturnType())
-      Val = Builder.CreatePointerBitCastOrAddrSpaceCast(
```
I couldn't break it with something like:
```
using int_as1 = int __attribute((address_space(1)));
using int_as2 = int __attribute((address_space(2)));

extern int_as1 foo();

//thread_local const int_as2 &r = foo(); // error
//thread_local const int_as1 &r = foo(); // error (?)
thread_local const int & __attribute((address_space(1))) r = foo(); // compiles (?)
```
The behavior is not intuitive to me, however.

JOE1994 replied:
Maybe I should play safe by just replacing the call to `CreateAddrSpaceCast` instead of stripping it entirely.

s-barannikov replied:
I'm in favor of stripping it. In the end, codebase simplification is the whole point of opaque pointers.
This check was added in https://reviews.llvm.org/D5353, which has nothing to do with address spaces.
If it ever breaks, it should be easy to fix.

s-barannikov reviewed (approved):
LGTM with the suggestion applied.

JOE1994 committed:
Update comment for CreatePointerBitCastOrAddrSpaceCast

JOE1994 committed:
git clang-format

bjope commented on clang/lib/CodeGen/CodeGenFunction.cpp:
```diff
@@ -1123,9 +1123,7 @@ void CodeGenFunction::StartFunction(GlobalDecl GD, QualType RetTy,
-      Builder.CreateStore(Builder.CreatePointerBitCastOrAddrSpaceCast(
-                              ReturnValue.getPointer(), Int8PtrTy),
-                          ReturnValuePointer);
+      Builder.CreateStore(ReturnValue.getPointer(), ReturnValuePointer);
```
So `ReturnValue.getPointer()` is guaranteed to be an unqualified pointer?

I'm just wondering why this hasn't been just a pointer bitcast in the past. But even if it would cast away an address space it seems a bit weird.
Not sure if it is worth an assert now when we bypass all the castIsValid checks etc that would have been done in the past.

(And maybe the "result.ptr" should be created with `ReturnValue.getPointer().getType()` rather than using Int8PtrTy. Although I consider that as out of scope for this patch.)

JOE1994 replied:
> But even if it would cast away an address space it seems a bit weird.

Yes, I agree..

> (And maybe the "result.ptr" should be created with ReturnValue.getPointer().getType() rather than using Int8PtrTy. Although I consider that as out of scope for this patch.)

I also agree.
Since, adding the change to this revision doesn't add new failures with `ninja check-clang`, I think it can be merged to this patch. I'll add the change to this revision.

JOE1994 committed:
Update type of 'result.ptr'

Assuming that 'ReturnValue.getPOinter()' is always from addrspace 0 doesn't seem right?
